Fix `ldelem[a]/stelem/cpobj/ldobj/stobj <pointer type>` import

SingleAccretion opened the pull request:
To treat pointers as integers and not `object`s.

Fixes #71636.

SingleAccretion committed:
Fix the ldelema case

SingleAccretion committed:
Add a test for the ldelem case

SingleAccretion committed:
Fix the ldelem case

SingleAccretion committed:
Add a test for the ldelema case

jkotas commented on src/coreclr/jit/importer.cpp:
```diff
@@ -13675,53 +13675,37 @@ void Compiler::impImportBlockCode(BasicBlock* block)
+            break;
 
             // ldelem for reference and value types
             case CEE_LDELEM:
```
`CEE_STELEM` has the same problem

SingleAccretion replied:
Indeed, and a couple other cases. I've fixed all of them.

jkotas commented on src/coreclr/jit/importer.cpp:
```diff
@@ -13675,53 +13675,37 @@ void Compiler::impImportBlockCode(BasicBlock* block)
-                // If it's a value class array we just do a simple address-of
-                if (eeIsValueClass(ldelemClsHnd))
+                // If it's a value class / pointer array, or a readonly access, we don't need a type check.
+                if ((lclTyp != TYP_REF) || ((prefixFlags & PREFIX_READONLY) != 0))
```
We can also apply `impCanSkipCovariantStoreCheck` to skip the helper, similar to how we do it in `CEE_STELEM_REF`.

For example, we can pass modify `impCanSkipCovariantStoreCheck` to treat `NULL` `value` as the ldelema case.

SingleAccretion committed:
Add a test for the stelem case

SingleAccretion committed:
Fix the stelem case

SingleAccretion committed:
Delete dead code while we're here

SingleAccretion committed:
Add a test for the cpobj case

SingleAccretion committed:
Fix the cpobj case

SingleAccretion committed:
Fix stobj for consistency

There was no actual problem with it because the "stelem.ref"
logic takes the type to store from the value to store, but
let's make the code obviously correct.

SingleAccretion committed:
Add a test for the ldobj case

SingleAccretion committed:
Fix the ldobj case

SingleAccretion committed:
Tighten asserts

They were relaxed in #50749.

SingleAccretion committed:
Add a TODO-CQ

EgorBo commented:
Thanks!
